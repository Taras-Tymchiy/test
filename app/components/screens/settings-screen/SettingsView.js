// @flow
import * as React from 'react';
import {
  Platform,
  Text,
  TextInput,
  View
} from 'react-native';
import FormControlWrapper from '../../molecules/FormControlWrapper';
import styles from './SettingsView.styles';

type SettingsViewState = {
  syncInterval: string;
  postsUrl: string;
  postsCount: string;
}

export type SettingsViewProps = {
  syncInterval: number;
  postsUrl: string;
  postsCount: number;
  intervalChanged: (v: number) => void;
  urlChanged: (v: string) => void;
  postsCountChanged: (v: number) => void;
}

export default class SettingsView 
  extends React.Component<SettingsViewProps, SettingsViewState> {
    constructor(props: SettingsViewProps) {
      super(props);
      this.state = this.propsToState(props);
    }

    propsToState({ syncInterval, postsUrl, postsCount }: SettingsViewProps) {
      return  { 
        postsUrl, 
        syncInterval: String(syncInterval),
        postsCount: String(postsCount)
      };
    }

    componentWillReceiveProps(newProps: SettingsViewProps) {
      this.setState(this.propsToState(newProps));
    }

    toNumeric = (val: string)=> String(val).replace(/[^0-9]/g, ''); 

    onIntervalChange = (v: string) => { this.setState({syncInterval: this.toNumeric(v)}) };
    onIntervalSave = () => { this.props.intervalChanged(+this.state.syncInterval) };

    onPostsCountChange = (v: string) => { this.setState({postsCount: this.toNumeric(v)}) };
    onPostsCountSave = () => { this.props.postsCountChanged(+(this.state.postsCount || 0)) };
    
    onUrlChange = (postsUrl: string) => { this.setState({postsUrl}) };
    onUrlSave = () => { this.props.urlChanged(this.state.postsUrl) };
    
    onDiscard = () => {
      this.setState(this.propsToState(this.props)); 
    };
    
    render () {
      const { syncInterval, postsUrl, postsCount } = this.state;
      return (
        <View style={styles.container} >
          {/* $FlowFixMe */}
          <FormControlWrapper 
              title="Endpoint url"
              value={postsUrl} 
              onSave={this.onUrlSave} 
              onDiscard={this.onDiscard} >
            <TextInput value={postsUrl} multiline={true} onChangeText={this.onUrlChange} />
          </FormControlWrapper>
          <FormControlWrapper 
              title="Posts number"
              description="The number of posts to fetch from API and display (0 means unlimited)"
              value={postsCount} 
              onSave={this.onPostsCountSave}
              keyboardType = 'numeric'
              onDiscard={this.onDiscard} >
            <TextInput value={postsCount} onChangeText={this.onPostsCountChange} />
          </FormControlWrapper>
          <FormControlWrapper 
              title="Sync interval" 
              description="Synchronization interval in milliseconds. Values less than 1000 are not allowed and will be overwritten to 1000"
              value={`${syncInterval} ms`} 
              onSave={this.onIntervalSave} 
              keyboardType = 'numeric'
              onDiscard={this.onDiscard} >
            <TextInput value={syncInterval} onChangeText={this.onIntervalChange} />
          </FormControlWrapper>
        </View>
      );
    }
};
