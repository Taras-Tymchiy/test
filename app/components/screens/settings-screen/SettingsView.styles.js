// @flow
import { StyleSheet } from 'react-native';
import { SPACING_L, COLOR_ELEMENT_BORDER } from '../../styles/constants';

export default StyleSheet.create({
  container: {
    marginTop: SPACING_L,
    marginBottom: SPACING_L,
    borderTopWidth: StyleSheet.hairlineWidth,
    borderColor: COLOR_ELEMENT_BORDER,
  }
});