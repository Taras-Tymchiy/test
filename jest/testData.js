import { Post } from "../app/entities/Post";


export const posts: Post[] = [
    {
        id: 1,
        id_str: '1',
        created_at: '2015-01-01',
        user: {
            id: 1,
            name: 'John Snow',
        },
        text: 'post text 1'
    },
    {
        id: 2,
        id_str: '2',
        created_at: '2015-01-02',
        user: {
            id: 2,
            name: 'Barack Obama',
        },
        text: 'post text 2'
    }
]