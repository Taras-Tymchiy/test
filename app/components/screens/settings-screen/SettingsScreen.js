// @flow
import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import * as settingsActions from '../../../store/actions/SettingsActions';
import { type RootState } from '../../../store/reducers/StateTypes';
import SettingsView from './SettingsView';

interface SettingsScreenProps {
  +pullInterval: number;
  +postsCount: number;
  +postsUrl: string;
  +actions: typeof settingsActions;
}

export class SettingsScreen extends Component<SettingsScreenProps> {
  render() {
    const { pullInterval, postsUrl, actions, postsCount } = this.props;
    return (
      <SettingsView 
        pullInterval={pullInterval}
        postsUrl={postsUrl}
        postsCount={postsCount}
        intervalChanged={this.updateInterval}
        urlChanged={this.updateUrl}
        postsCountChanged={this.updatePostsCount}
      />
    );
  }

  updateInterval = (v: number) => {
    this.props.actions.updateInterval(v);
  }

  updateUrl = (v: string) => {
    this.props.actions.updateUrl(v);
  }
  
  updatePostsCount = (v: number) => {
    this.props.actions.updatePostsCount(v);
  }
}

export default connect(
  ({settings: {pullInterval, postsUrl, postsCount}}: RootState) => ({pullInterval, postsUrl, postsCount}),
  dispatch => ({actions: bindActionCreators(settingsActions, dispatch)})
)(SettingsScreen);
