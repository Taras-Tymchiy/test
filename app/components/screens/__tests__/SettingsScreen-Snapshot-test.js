import React from 'react';
import renderer from 'react-test-renderer';
import { SettingsScreen } from '../settings-screen/SettingsScreen';
import { posts } from '../../../../jest/testData';

jest.mock('../settings-screen/SettingsView', (props) => `SettingsView ${props}`);

const props = {
  pullIntervl: 5000,
  postsCount: 10,
  postsUrl: 'test',
  actions: {
    updateInterval: () => null,
    updateUrl: () => null,
    updatePostsCount: () => null
  }
};

it('renders posts in a SettingsScreen using Snapshots', () => {
  expect(renderer.create(
    <SettingsScreen {...props} />
  )).toMatchSnapshot();
});
