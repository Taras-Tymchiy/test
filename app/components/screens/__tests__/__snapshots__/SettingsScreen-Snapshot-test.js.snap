// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders posts in a SettingsScreen using Snapshots 1`] = `
<SettingsView undefined
  intervalChanged={[Function]}
  postsCount={10}
  postsCountChanged={[Function]}
  postsUrl="test"
  pullInterval={undefined}
  urlChanged={[Function]}
/>
`;
